cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -133,31 +215,111 @@ class MatrixCSR
 
   /// Insertion functor for setting values in matrix. It is typically
   /// used in finite element assembly functions.
-  /// @param A Matrix to insert into
-  /// @return Function for inserting values into `A`
+  /// @return Function for inserting values into the matrix
   /// @todo clarify setting on non-owned enrties
-  auto mat_set_values()
+  std::function<int(const xtl::span<const std::int32_t>& rows,
```

Comment:
This should be changed back to `auto` to allow inlining of the anonymous function. Will need redesign of the function body.

Reply:
redesign? Not sure how to. If I put `auto` then I get `inconsistent deduction for auto return type`.

Reply:
```
inconsistent deduction for auto return type: 
‘dolfinx::la::MatrixCSR<T, Allocator>::mat_add_values() [with T = double; Allocator = std::allocator<double>]::<lambda(const tcb::span<const int>&, const tcb::span<const int>&, const tcb::span<const double>&)>’ 
and then 
‘dolfinx::la::MatrixCSR<T, Allocator>::mat_add_values() [with T = double; Allocator = std::allocator<double>]::<lambda(const tcb::span<const int>&, const tcb::span<const int>&, const tcb::span<const double>&)>’
```
which doesn't really make sense, as it says the two identical types are inconsistent. I suppose it really treats each lambda function as a "type" so you can only return one actual lambda function with `auto`, and not choose which one to return.

Reply:
I don't think it's possible to have auto there in C++ 17.  Every lambda has a unique type and it's not defined at compile time. 
One alternative I can think of is to template the Matrix<T, BS>. So if BS is known at compile time we can use `cosntexpr if`.